src/query/functions/src/scalars/decimal.rs

```diff
@@ -468,13 +473,72 @@ fn convert_to_decimal(
     dest_type: DataType,
 ) -> Value<AnyType> {
     let arg = &args[0];
-    if from_type.is_integer() {
-        return integer_to_decimal(arg, ctx, from_type, dest_type);
+    match from_type {
+        DataType::Number(ty) => {
+            if ty.is_float() {
+                float_to_decimal(arg, ctx, from_type, dest_type)
+            } else {
+                integer_to_decimal(arg, ctx, from_type, dest_type)
+            }
+        }
+        DataType::Decimal(_) => decimal_to_decimal(arg, ctx, from_type, dest_type),
+        DataType::String => string_to_decimal(arg, ctx, from_type, dest_type),
+        _ => unreachable!("to_decimal not support this DataType"),
     }
-    if from_type.is_floating() {
-        return float_to_decimal(arg, ctx, from_type, dest_type);
+}
+
+fn string_to_decimal_internal<T: Decimal>(
+    ctx: &mut EvalContext,
+    string_column: &StringColumn,
+    size: DecimalSize,
+    dest_type: &DecimalDataType,
+) -> DecimalColumn {
+    let mut builder = DecimalDeserializer::<T>::with_capacity(dest_type, string_column.len());
+    for (row, buf) in string_column.iter().enumerate() {
+        match read_decimal_with_size::<T>(buf, size, true) {
+            Ok((d, _)) => builder.values.push(d),
+            Err(e) => {
+                ctx.set_error(row, format!("fail to decode string as decimal: {e}"));
+                builder.values.push(T::zero())
+            }
+        }
+    }
+    T::to_column(std::mem::take(&mut builder.values), size)
+}
+
+fn string_to_decimal(
+    arg: &ValueRef<AnyType>,
+    ctx: &mut EvalContext,
+    from_type: DataType,
+    dest_type: DataType,
+) -> Value<AnyType> {
+    let dest_type = dest_type.as_decimal().unwrap();
+
+    let mut is_scalar = false;
+    let column = match arg {
+        ValueRef::Column(column) => column.clone(),
+        ValueRef::Scalar(s) => {
+            is_scalar = true;
+            let builder = ColumnBuilder::repeat(s, 1, &from_type);
```

Comment:
Don't need to convert `Scalar` to a `Column`, we can get the value from both `Scalar` and `Column` by function `index_unchecked` or `index`.

Reply:
you are right,  I just followed the impl of  prev similar functions. 
I will change it.

Reply:
I think we mainly optimize of the Column( not the Scalar) case.
it is slower to braching for each iterm?

```
    pub unsafe fn index_unchecked(&'a self, index: usize) -> T::ScalarRef<'a> {
        match self {
            ValueRef::Scalar(scalar) => scalar.clone(),
            ValueRef::Column(c) => T::index_column_unchecked(c, index),
        }
    }
``

Reply:
`Scalar` only has one item, so it's ok to clone. You can refer to the implementation of this function.
https://github.com/datafuselabs/databend/blob/main/src/query/functions/src/scalars/array.rs#L121

Reply:
`Scalar` only has one item, so it's ok to clone. You can refer to the implementation of this function.
https://github.com/datafuselabs/databend/blob/main/src/query/functions/src/scalars/array.rs#L121

Reply:
not the clone of scala, I mean the `match`

Reply:
I refactored with a more straightforward way.